Pipelines support testing prs from forks

scottwittenburg opened the pull request:
Changes to support running pipelines on pull requests from forks.  At a high level, this change supports:

- Running pipelines with no secrets whatsoever, or whatever secrets do happen to be available 
- A code path for PRs allowing pushing binaries to an alternate mirror.  This will improve developer experience in cases when the first push to a PR branch causes a rebuild to an expensive package, but subsequent pushes wouldn't cause subsequent rebuilds.
- Makes use of #19209 so that only a single install pass is required to install packages (spack would previously have installed a binary cache even if it's `full_hash` was different than the locally computed one).  Allows us to do away with:
  * Do a first pass install with `--only dependencies` and `--cache-only`
  * Replace the re-written `spack.yaml` which was rendered useless by the first install
  * Do a second pass install with `--only package` and `--no-cache`

The changes here are based on #19209 and require that to be merged first.

scottwittenburg committed:
Pipelines: Support running gitlab pipelines on PRs from forks

This change makes improvements to the `spack ci rebuild` command
which supports running gitlab pipelines on PRs from forks.  Much
of this has to do with making sure we can run without the secrets
previously required for running gitlab pipelines (e.g signing key,
aws credentials, etc).  Specific improvements in this PR:

Check if spack has precisely one signing key, and use that information
as an additional constraint on whether or not we should attempt to sign
the binary package we create.

Also, if spack does not have at least one public key, add the install
option "--no-check-signature"

If we are running a pipeline without any profile or environment
variables allowing us to push to S3, the pipeline could still
successfully create a buildcache in the artifacts and move on.  So
just print a message and move on if pushing either the buildcache
entry or cdash id file to the remote mirror fails.

When we attempt to generate a pacakge or gpg key index on an S3
mirror, and there is nothing to index, just print a warning and
exit gracefully rather than throw an exception.

Support the use of PR-specific mirrors for temporary binary pkg
storage.  This will allow quality-of-life improvement for developers,
providing a place to store binaries over the lifetime of a PR, so
that they must only wait for packages to rebuild from source when
they push a new commit that causes it to be necessary.

Replace two-pass install with a single pass and the new option:
 --require-full-hash-match.  Doing this also removes the need to
save a copy of the spack.yaml to be copied over the one spack
rewrites in between the two spack install passes.

Work around a mirror configuration issue caused by using
spack.util.executable to do the package installation.

scottwittenburg commented:
@scheibelp Depending on how testing goes the next day or so, I may need to make a few changes, but I think this is ready for at least a first round of review.

opadron commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -703,10 +703,23 @@ def generate_package_index(cache_prefix):
+            entry
+            for entry in web_util.list_url(cache_prefix)
+            if entry.endswith('.yaml'))
+    except KeyError as inst:
```
Is this the same `KeyError` that we talked about from the s3 implementation of `list_url`?  I thought we had addressed this error?

scottwittenburg replied:
It's only addressed here, as far as I know, unless you got a fix merged at some point?

opadron commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -768,10 +781,23 @@ def generate_key_index(key_prefix, tmpdir=None):
+            entry[:-4]
+            for entry in web_util.list_url(key_prefix, recursive=False)
+            if entry.endswith('.pub'))
+    except KeyError as inst:
```
Same question about this `KeyError`.

opadron commented on lib/spack/spack/cmd/ci.py:
```diff
@@ -304,82 +347,63 @@ def ci_rebuild(args):
+                # that exception type here, since users of the
+                # `spack ci rebuild' may not need or want any dependency
+                # on boto3.  So we use the first non-boto exception type
+                # in the heirarchy:
```
I wonder if it makes sense to shim out the few Exception classes needed from boto?  I'll leave it up to you to decide if the maintenance cost is worth it.

scottwittenburg replied:
I'm not sure it's worth the maintenance cost, but I could be convinced, maybe I'm just not seeing the value.

opadron reviewed (changes requested):
Wondering what the situation is with those `KeyErrors`.  Otherwise, LGTM.

scottwittenburg committed:
Update pipeline trigger jobs for PRs from forks

Moving to PRs from forks relies on external synchronization script
pushing special branch names.  Also secrets will only live on the
spack mirror project, and must be propagated to the E4S project via
variables on the trigger jobs.

When this change is merged, pipelines will not run until we update
the "Custom CI configuration path" in the Gitlab CI Settings, as the
name of the file has changed to better reflect its purpose.

scottwittenburg committed:
Arg to MirrorCollection is used exclusively, so add main remote mirror to it

scottwittenburg committed:
Compute full hash less frequently

scottwittenburg committed:
Add tests covering index generation error handling code

opadron reviewed (approved):
With respect to the `KeyError`, I'm pretty sure that the issue is on/around this line:

https://github.com/spack/spack/blob/6021ac00af6e4adc8a520cbb4b0f53330e7f4176/lib/spack/spack/util/web.py#L275

If you'd like, feel free to take a stab at moving your fix for this from catching the `KeyError` in your code to that part of the code.

Alternatively, if you could make a new issue with a link to this comment, and assign it to me, I'll fix it after the fact (eventually...).

scottwittenburg commented:
@scheibelp IIUC @tgamblin hopes to merge this by Wednesday this week.  Do you think you'll have time to review it sometime today?  I'll try to respond to your review suggestions immediately.  Let me know if you want to sync up verbally for some background.

tgamblin reviewed (approved):
@scottwittenburg: this looks good for the moment but we should get to the bottom of the Exceptions that we're squashing.  The error messages should help with that.

alalazo commented:
@tgamblin Since this PR has been merged I'm not seeing any notification in `develop` for the e4s pipeline. Is this expected?

Here's the last commit that shows it:
![Screenshot from 2020-11-18 14-11-36](https://user-images.githubusercontent.com/4199709/99534770-3c6db780-29a8-11eb-8693-e04a3ceafe0c.png)

scottwittenburg commented:
@alalazo We are in the process of "flipping some switches" in gitlab to support running PRs from forks, and while we're ironing out the wrinkles, I have disabled gitlab pipelines.  I expect we'll have them running again in the next day or so.

alalazo commented:
@scottwittenburg Thank you for the info. Looking forward to have gitlab pipelines in PRs!
